### PR TITLE
ENH: Adding engine_kwargs to Excel engines for issue #40274

### LINKED ISSUE #40274: ENH: read_excel (xlrd engine) add parameter for ignore_workbook_corruption
#### Is your feature request related to a problem?

yes, sometimes, expecially when created by 3rd party application old excel file can be corrupted:

https://www.codeforests.com/2020/05/28/fix-the-compdocerror-for-xlrd/
https://stackoverflow.com/questions/12705527/reading-excel-files-with-xlrd
https://stackoverflow.com/questions/34550624/python-xlrd-read-excel-file-error/
https://titanwolf.org/Network/Articles/Article?AID=f1370e16-c9ef-4892-a00e-b1bafee2f6a8#gsc.tab=0

#### Describe the solution you'd like

xlrd's team recently in version 2.0 (december 2020) solved this with a flag: `ignore_workbook_corruption`, please see https://xlrd.readthedocs.io/en/latest/changes.html
So my desire would be to just have this flag inside the pandas's function `read_excel`

#### API breaking implications

any, I guess

#### Describe alternatives you've considered

Open the file directly with xlrd and pass it to pandas
I may be unaware of other solutions

### PR DESCRIPTION
**Pull request by rmhowe425**
- [X] closes #40274 (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by rmhowe425** [comment]
@rhshadrach This PR should be ready to go! 

Please let me know if the PR requires anything else!

File: pandas/tests/io/excel/test_readers.py [hunk 1]
```diff
@@ -148,6 +148,23 @@ def parser(self, *args, **kwargs):
...
+            "ods": {},
+        }
+
+        df = pd.read_excel(
+            "test1" + read_ext,
+            sheet_name="Sheet1",
+            index_col=0,
+            **expected_defaults[read_ext[1:]],
+        )
+        assert not df.empty
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:166]: I don't think this is a sufficient test. Can you instead pass an invalid argument (e.g. `foo`) and check that the expected exception is raised.

File: pandas/tests/io/excel/test_readers.py [hunk 2]
```diff
@@ -148,6 +148,23 @@ def parser(self, *args, **kwargs):
             expected = expected_defaults[read_ext[1:]]
         assert result == expected
 
+    def test_engine_kwargs(self, read_ext):
+        expected_defaults = {
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:152]: Can you add `# GH#52214` as the first line of the test

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the PR!

Ref: https://github.com/pandas-dev/pandas/issues/40274#issuecomment-1467204047

I think we need to add documentation on which function pandas is using for each engine. I'd recommend adding it to the User Guide and then adding a link to this in the docstring.

Otherwise, generally looks good!

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:166]: @rhshadrach when you say "document which function pandas is using for each engine" do you mean that we should be documenting the class associated with each engine? For example, if I want to use pandas.read_excel to read an .xlrd file, I would be using XlrdReader.load_workbook to read in the file?

Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:167]: @rhshadrach when you say "document which function pandas is using for each engine" do you mean that we should be documenting the class associated with each engine? For example, if I want to use pandas.read_excel to read an .xlrd file, I would be using XlrdReader.load_workbook to read in the file.

Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py]: @rhshadrach Would you be able to provide any insight into why `TestMySQLAlchemy.test_to_sql_with_negative_npinf` is failing in all of the Ubuntu tests? That test is skipped when I run it locally on my Ubuntu systemusing PyTest.

image


Also, the failing Doc Build isn't giving me any insight into the error either which is strange

Comment (rhshadrach) [reply pandas/tests/io/excel/test_readers.py]: The SQL test failure looks unrelated; the docs build is failing because of a warning - also looks unrelated.

File: doc/source/user_guide/io.rst [hunk 3]
```diff
@@ -3449,6 +3449,14 @@ Reading Excel files
 In the most basic use-case, ``read_excel`` takes a path to an Excel
 file, and the ``sheet_name`` indicating which sheet to parse.
 
+* When reading in (``.xlsx``) and (``.xlsm``) files, pandas is using :func:`~openpyxl.load_workbook`.
```
Comment (rhshadrach) [inline doc/source/user_guide/io.rst:3452]: The function used is dependent on the engine and not the extension. Even if the relation `extension ~ engine` is well-defined today, it may not be in the future. Can you rephrase as something like:

> For the engine `openpyxl`, pandas is using...

Also, I think it'd be good to mention _why_ we're telling the user this. Something like "When using engine_kwargs, pandas will pass arguments through to the engine. For this, it is important to know which function pandas is using internally."

File: doc/source/whatsnew/v2.1.0.rst [hunk 4]
```diff
@@ -38,6 +38,7 @@ Other enhancements
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+- :meth:`pandas.read_excel` now supports ``engine_kwargs`` to pass keyword args to Excel engines. (:issue:`40274`)
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:41]: nit: no period at the end (although there are other violations of this in this particular section)

File: pandas/io/excel/_base.py [hunk 5]
```diff
@@ -289,6 +289,8 @@
 
     .. versionadded:: 2.0
 
+engine_kwargs : Arbitrary keyword arguments passed to excel engine
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:292]: This line should only have the type and default (if applicable). The line after has the description. Also add a `versionadded`.

File: pandas/io/excel/_base.py [hunk 6]
```diff
@@ -302,6 +304,11 @@
 read_csv : Read a comma-separated values (csv) file into DataFrame.
 read_fwf : Read a table of fixed-width formatted lines into DataFrame.
 
+Notes
+-----
+For specific information on the methods used for each Excel engine, refer to the pandas
+user guide: https://pandas.pydata.org/docs/user_guide/io.html#excel-files.
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:310]: This is an internal cross reference, don't use an absolute URL.

https://docs.readthedocs.io/en/stable/guides/cross-referencing-with-sphinx.html

File: pandas/io/excel/_base.py [hunk 7]
```diff
@@ -472,13 +479,15 @@ def read_excel(
     skipfooter: int = 0,
     storage_options: StorageOptions = None,
     dtype_backend: DtypeBackend | lib.NoDefault = lib.no_default,
+    **engine_kwargs,
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:482]: This is currently inconsistent with the use of engine_kwargs elsewhere in pandas (excluding SQL). Use `engine_kwargs`, not `**engine_kwargs`.

File: pandas/tests/io/excel/test_readers.py [hunk 8]
```diff
@@ -148,6 +149,53 @@ def parser(self, *args, **kwargs):
...
+            "xlsx": {"foo": "abcd"},
+            "xlsm": {"foo": 123},
+            "xlsb": {"foo": "True"},
+            "xls": {"foo": True},
+            "ods": {},
+        }
+
+        # odf.opendocument does not take any additional args
+        if read_ext[1:] != "ods":
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:182]: Why are we excluding ods here?

File: pandas/tests/io/excel/test_readers.py [hunk 9]
```diff
@@ -148,6 +149,53 @@ def parser(self, *args, **kwargs):
             expected = expected_defaults[read_ext[1:]]
         assert result == expected
 
+    def test_engine_kwargs_valid(self, read_ext, engine):
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:152]: Does this test cover something the other test doesn't?

File: pandas/io/excel/_base.py (hunk 5 above)
Comment (rhshadrach) [reply pandas/io/excel/_base.py:292]: Note this comment also applies to other docstrings

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the changes! There were a few things I missed on the first pass.

File: pandas/tests/io/excel/test_readers.py (hunk 1 above)
Comment (rhshadrach) [reply pandas/tests/io/excel/test_readers.py]: > If the SQL test failure and doc build failures are unrelated is there anything left for me to do?

Not in this regard, no. It is okay to ignore unrelated failures.

File: pandas/tests/io/excel/test_readers.py (hunk 8 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:182]: @rhshadrach The `load` method used to read in .ods in for _odfreader.py takes 1 and only 1 argument

File: pandas/tests/io/excel/test_readers.py (hunk 9 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:152]: @rhshadrach 
The test_engine_kwargs_valid function tests valid arguments being passed to the load_workbook method for each engine.

The invalid method tests args that are not supported the various engines

File: pandas/tests/io/excel/test_readers.py (hunk 8 above)
Comment (rhshadrach) [reply pandas/tests/io/excel/test_readers.py:182]: I see, we should still be passing engine_kwargs through. It may not in the future, and if a user tries to pass engine_kwargs today with that reader, their code should fail.

File: pandas/tests/io/excel/test_readers.py (hunk 9 above)
Comment (rhshadrach) [reply pandas/tests/io/excel/test_readers.py:152]: > The test_engine_kwargs_valid function tests valid arguments being passed to the load_workbook method for each engine.

It doesn't do this. It only tests that the call returns a non-empty frame. This test will still pass even if we didn't pass the arguments to the engine.

File: pandas/tests/io/excel/test_readers.py (hunk 8 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:182]: Understood! Will fix this!

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3452]: @rhshadrach Would this be acceptable: `For the engine openpyxl, pandas is using :func:`~openpyxl.load_workbook` to read in xlsx and xlsm files.`

**Comment by rmhowe425** [comment]
@rhshadrach Changes have been made! Please let me know if I have missed anything else!

**Comment by rmhowe425** [comment]
Fixed a merge conflict with `doc/source/whatsnew/v2.1.0.rst`

**Comment by rmhowe425** [comment]
@rhshadrach Wondering if I should submit a GH issue to implement the same feature for the `to_excel` method?

File: pandas/io/excel/_base.py [hunk 10]
```diff
@@ -520,7 +533,10 @@ def read_excel(
 
 class BaseExcelReader(metaclass=abc.ABCMeta):
     def __init__(
-        self, filepath_or_buffer, storage_options: StorageOptions = None
+        self,
+        filepath_or_buffer,
+        storage_options: StorageOptions = None,
+        **engine_kwargs,
```
Comment (rhshadrach) [inline pandas/io/excel/_base.py:539]: `engine_kwargs` should be treated consistently throughout - can you make this `engine_kwargs` instead of `**engine_kwargs` throughout

File: pandas/io/excel/_odfreader.py [hunk 11]
```diff
@@ -40,20 +41,25 @@ def __init__(
         filepath_or_buffer : str, path to be parsed or
             an open readable stream.
         {storage_options}
+        engine_kwargs : Arbitrary keyword arguments passed to excel engine
```
Comment (rhshadrach) [inline pandas/io/excel/_odfreader.py:44]: This an the rest of the docstrings need to be corrected - the first line contains the type and default, the second line and after contains the descriptions.

File: pandas/tests/io/excel/test_readers.py [hunk 12]
```diff
@@ -148,6 +149,26 @@ def parser(self, *args, **kwargs):
...
+        # GH#52214
+        expected_defaults = {
+            "xlsx": {"foo": "abcd"},
+            "xlsm": {"foo": 123},
+            "xlsb": {"foo": "True"},
+            "xls": {"foo": True},
+            "ods": {},
+        }
+
+        msg = re.escape(r"read_excel() got an unexpected keyword argument 'foo'")
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:162]: This is not the expected error message - it should be when pandas calls the engine with an incorrect kwarg that should raise. Below, the line should be `engine_kwargs=expected_defaults[...]`.

**Review by rhshadrach — CHANGES_REQUESTED** [review]

**Comment by rhshadrach** [comment]
> @rhshadrach Wondering if I should submit a GH issue to implement the same feature for the `to_excel` method?

Yea - I think that makes sense to do!

File: doc/source/user_guide/io.rst [hunk 13]
```diff
@@ -3449,6 +3449,18 @@ Reading Excel files
 In the most basic use-case, ``read_excel`` takes a path to an Excel
 file, and the ``sheet_name`` indicating which sheet to parse.
 
+When using the engine_kwargs parameter, pandas will pass these arguments to the
+appropriate engine. For this, it is important to know which function pandas is
+using internally.
+
+* For the engine openpyxl, pandas is using :func:`~openpyxl.load_workbook` to read in (``.xlsx``) and (``.xlsm``) files.
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3456]: @rhshadrach Sorry, I don't quite understand. We already have `openpyxl.load_workbook` above. How would we change that?

File: doc/source/whatsnew/v2.1.0.rst [hunk 14]
```diff
@@ -41,10 +41,11 @@ Other enhancements
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+
+- :meth:`pandas.read_excel` now supports ``engine_kwargs`` to pass keyword args to Excel engines. (:issue:`40274`).
+- Add dtype of categories to ``repr`` information of :class:`CategoricalDtype` (:issue:`52179`).
 - :meth:`Categorical.map` and :meth:`CategoricalIndex.map` now have a ``na_action`` parameter (:issue:`44279`)
-- Add dtype of categories to ``repr`` information of :class:`CategoricalDtype` (:issue:`52179`)
 -
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:48]: Yeah this came from handling a merge conflict. Both highlighted lines are the same. I moved the line causing the merge conflict into my new push.

File: doc/source/user_guide/io.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply doc/source/user_guide/io.rst:3456]: The reference ``:func:`~openpyxl.load_workbook` `` will display as `load_workbook`. I'm suggesting to do ``:func:`openpyxl.load_workbook` `` which will display as `openpyxl.load_workbook` instead.

[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3456]: ah! Understood!

File: doc/source/user_guide/io.rst [hunk 15]
```diff
@@ -3449,6 +3449,18 @@ Reading Excel files
 In the most basic use-case, ``read_excel`` takes a path to an Excel
 file, and the ``sheet_name`` indicating which sheet to parse.
 
+When using the engine_kwargs parameter, pandas will pass these arguments to the
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3452]: Completed this change

File: doc/source/user_guide/io.rst [hunk 16]
```diff
@@ -3449,6 +3449,18 @@ Reading Excel files
...
+appropriate engine. For this, it is important to know which function pandas is
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3453]: Completed this change

File: doc/source/user_guide/io.rst (hunk 13 above)
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst:3456]: Completed this change

File: doc/source/whatsnew/v2.1.0.rst [hunk 17]
```diff
@@ -41,10 +41,11 @@ Other enhancements
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:44]: Completed this change

**Comment by rhshadrach** [comment]
@rmhowe425 Please be aware that each time you merge main into this branch, it kicks off all the tests (30+hours of compute). This should be done only when needed - for example:
 - if there are conflicts
 - if you're aware of a PR that has been merged to main that touches on similar parts of the code
 - right before we merge this branch into main (to make sure all the tests still pass)
I don't think it needs to be done daily, or even once every couple of days.

File: doc/source/user_guide/io.rst [hunk 18]
```diff
@@ -3449,6 +3449,18 @@ Reading Excel files
...
+using internally.
+
+* For the engine openpyxl, pandas is using :func:`~openpyxl.load_workbook` to read in (``.xlsx``) and (``.xlsm``) files.
+
+* For the engine xlrd, pandas is using :func:`~xlrd.open_workbook` to read in (``.xls``) files.
+
+* For the engine xlrd, pandas is using :func:`~pyxlsb.open_workbook` to read in (``.xlsb``) files.
```
[the remark this answers is not in the document]
Comment (rhshadrach) [reply doc/source/user_guide/io.rst]: This has not been resolved.

File: pandas/io/excel/_base.py (hunk 10 above)
Comment (rmhowe425) [reply pandas/io/excel/_base.py:539]: Done!

File: pandas/io/excel/_odfreader.py (hunk 11 above)
Comment (rmhowe425) [reply pandas/io/excel/_odfreader.py:44]: Done!

File: pandas/io/excel/_base.py [hunk 19]
```diff
@@ -472,13 +480,18 @@ def read_excel(
     skipfooter: int = 0,
     storage_options: StorageOptions = None,
     dtype_backend: DtypeBackend | lib.NoDefault = lib.no_default,
+    engine_kwargs: dict = dict(),
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply pandas/io/excel/_base.py]: Done!

File: doc/source/user_guide/io.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/user_guide/io.rst]: image
Line 3460 has been updated to include `pyxlsb`. Is there something else that I'm missing?

File: pandas/tests/io/excel/test_readers.py [hunk 20]
```diff
@@ -148,6 +149,30 @@ def parser(self, *args, **kwargs):
...
         assert result == expected
 
+    def test_engine_kwargs(self, read_ext, engine):
+        # GH#52214
+        expected_defaults = {
+            "xlsx": {"foo": "abcd"},
+            "xlsm": {"foo": 123},
+            "xlsb": {"foo": "True"},
+            "xls": {"foo": True},
+            "ods": {},
```
Comment (rhshadrach) [inline pandas/tests/io/excel/test_readers.py:159]: Can you add the args for ods - this currently isn't being tested.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Looking good! I think this is almost ready.

File: pandas/tests/io/excel/test_readers.py (hunk 20 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:159]: Done!

File: doc/source/whatsnew/v2.1.0.rst [hunk 21]
```diff
@@ -42,10 +42,10 @@ Other enhancements
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+- :meth:`pandas.read_excel` now supports ``engine_kwargs`` to pass keyword args to Excel engines. (:issue:`40274`).
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:45]: Done!

**Comment by rmhowe425** [comment]
@rhshadrach 
image
For some reason it seems like the Code Checks test isn't liking my attempts to revert this change

File: doc/source/whatsnew/v2.1.0.rst [hunk 22]
```diff
@@ -42,10 +42,9 @@ Other enhancements
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
 - :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
```
Comment (rhshadrach) [inline doc/source/whatsnew/v2.1.0.rst:44]: Can you remove the period at the end of this line? I believe is it messing up our pre-commit script.

Comment (rhshadrach) [reply doc/source/whatsnew/v2.1.0.rst:44]: cc @MarcoGorelli - just pinging for awareness.

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst]: @rhshadrach Just removed the period and pushed.

**Comment by rmhowe425** [comment]
Fixed merge conflict

File: doc/source/whatsnew/v2.1.0.rst [hunk 23]
```diff
@@ -85,10 +85,10 @@ Other enhancements
 - Added to the escape mode "latex-math" preserving without escaping all characters between "\(" and "\)" in formatter (:issue:`51903`)
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
 - :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
-- :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`).
+- :meth:`arrays.SparseArray.map` now supports ``na_action`` (:issue:`52096`)
 - Add dtype of categories to ``repr`` information of :class:`CategoricalDtype` (:issue:`52179`)
--
-
+- :meth:`pandas.read_excel` now supports ``engine_kwargs`` to pass keyword args to Excel engines. (:issue:`40274`)
+- :meth:`Categorical.map` and :meth:`CategoricalIndex.map` now have a ``na_action`` parameter (:issue:`44279`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.1.0.rst:91]: looks like this line is now duplicated, it also appears on line 74

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for your PR

Just got some comments

1. the docs build is failing with:

```
/home/runner/work/pandas/pandas/doc/source/whatsnew/v2.1.0.rst:106: WARNING: Bullet list ends without a blank line; unexpected unindent.
```
, you need to make sure there's a blank line before the `.. --------` line

2. regarding the code checks, could you please:
  1. remove the duplicate entry (each entry should only appear once)
  2. run `pre-commit run sort-whatsnew-items --all-files`
  3. stage, commit, push

File: doc/source/whatsnew/v2.1.0.rst [hunk 24]
```diff
@@ -84,10 +85,8 @@ Other enhancements
 - Improve error message when having incompatible columns using :meth:`DataFrame.merge` (:issue:`51861`)
 - Added to the escape mode "latex-math" preserving without escaping all characters between "\(" and "\)" in formatter (:issue:`51903`)
 - Improved error message when creating a DataFrame with empty data (0 rows), no index and an incorrect number of columns. (:issue:`52084`)
-- :meth:`DataFrame.applymap` now uses the :meth:`~api.extensions.ExtensionArray.map` method of underlying :class:`api.extensions.ExtensionArray` instances (:issue:`52219`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.1.0.rst:87]: why remove this one?

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:87]: Looking at the output from `pre-commit run sort-whatsnew-items --all-files` and the CI tests, I thought that was the one that I needed to remove. Removing it also greened up the above command when I ran it locally. 

I can add it back in the appropriate place

**Comment by rmhowe425** [comment]
@MarcoGorelli wow okay never mind. Not sure what I did wrong before....Thanks for pointing that mistake out to me!

**Comment by MarcoGorelli** [comment]
looks right now, nice :)

**Review by rhshadrach — APPROVED** [review]
lgtm, ping on green.

**Comment by rmhowe425** [comment]
@rhshadrach No idea why I'm still getting formatting errors for the whatsnew file. I've added a blank line.

**Comment by rmhowe425** [comment]
@MarcoGorelli PR should be good to go

**Review by rhshadrach — APPROVED** [review]

**Comment by rmhowe425** [comment]
@MarcoGorelli Looks like we're good to go! I think we just need your sign off?

File: doc/source/whatsnew/v2.1.0.rst [hunk 25]
```diff
@@ -71,9 +71,10 @@ to ``na_action=None``, like for all the other array types.
 
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
-- :meth:`Categorical.map` and :meth:`CategoricalIndex.map` now have a ``na_action`` parameter.
-  :meth:`Categorical.map` implicitly had a default value of ``"ignore"`` for ``na_action``. This has formally been deprecated and will be changed to ``None`` in the future.
+- :meth:`Categorical.map` implicitly had a default value of ``"ignore"`` for ``na_action``. This has formally been deprecated and will be changed to ``None`` in the future.
```
Comment (mroeschke) [inline doc/source/whatsnew/v2.1.0.rst:74]: Why are a lot of unrelated entries modified in this diff? I think only one entry should have been added

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:74]: @mroeschke Mistakes caused by handling merge conflicts. They should have been fixed from previous comments left by reviewers. Are we still seeing issues?

Comment (mroeschke) [reply doc/source/whatsnew/v2.1.0.rst:74]: Ideally for this pull request we should just see one new entry but other entries in this diff appear changed

**Comment by rmhowe425** [comment]
@mroeschke I'll pull the latest version of `whatsnew/v2.1.0.rst` from master, add my change for this PR and push. That should guarantee that all incorrect modifications to the file have been fixed

File: doc/source/whatsnew/v2.1.0.rst (hunk 25 above)
Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:74]: @mroeschke I'll pull the latest version of `whatsnew/v2.1.0.rst` from master, add my change for this PR and push. That should guarantee that all incorrect modifications to the file have been fixed

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst]: @mroeschke Updated `whatsnew/v2.1.0.rst` should be good

File: pandas/tests/io/excel/test_readers.py [hunk 26]
```diff
@@ -148,6 +149,33 @@ def parser(self, *args, **kwargs):
...
+            "ods": {"foo": "abcd"},
+        }
+
+        msg = re.escape(r"load_workbook() got an unexpected keyword argument 'foo'")
+
+        if read_ext[1:] == "xls" or read_ext[1:] == "xlsb":
+            msg = re.escape(r"open_workbook() got an unexpected keyword argument 'foo'")
+
+        elif read_ext[1:] == "ods":
+            msg = re.escape(r"load() got an unexpected keyword argument 'foo'")
```
Comment (MarcoGorelli) [inline pandas/tests/io/excel/test_readers.py:168]: the extra newlines makes this hard to read, how about
```python
if read_ext[1:] == "xls" or read_ext[1:] == "xlsb":
    msg = ...
elif read_ext[1:] == "ods":
    msg = ...
else:
    msg = ...
```
?

File: doc/source/whatsnew/v2.1.0.rst [hunk 27]
```diff
@@ -339,7 +340,6 @@ Period
 - Bug in :meth:`arrays.PeriodArray.map` and :meth:`PeriodIndex.map`, where the supplied callable operated array-wise instead of element-wise (:issue:`51977`)
 - Bug in :func:`read_csv` not processing empty strings as a null value, with ``engine="pyarrow"`` (:issue:`52087`)
 - Bug in :func:`read_csv` returning ``object`` dtype columns instead of ``float64`` dtype columns with ``engine="pyarrow"`` for columns that are all null with ``engine="pyarrow"`` (:issue:`52087`)
-- Bug in incorrectly allowing construction of :class:`Period` or :class:`PeriodDtype` with :class:`CustomBusinessDay` freq; use :class:`BusinessDay` instead (:issue:`52534`)
```
Comment (MarcoGorelli) [inline doc/source/whatsnew/v2.1.0.rst:342]: something went wrong when merging here, I presume you didn't mean to remove this line?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks @rmhowe425 

just left some comments

sorry for the conflicting reviews regarding removing the trailing period from the whatsnew note. if you wanted to handle that in a separate pull request, that would be welcome, but for this one, let's try to keep the diff minimal

File: doc/source/whatsnew/v2.1.0.rst (hunk 27 above)
Comment (MarcoGorelli) [reply doc/source/whatsnew/v2.1.0.rst:342]: > Shouldn't my PR only contain my contribution?

yes, exactly - whereas currently, it's showing that you also deleted another line

please check https://github.com/pandas-dev/pandas/pull/52214/files to verify that the PR only contains your changes

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:342]: @MarcoGorelli Looking at that URL, it looks correct to me. 

The PR only contains my changes. However, when I run into merge conflicts after other people's PRs are approved, do I need to add their changes when I examine the merge conflict diff, or remove them? In this case I removed them, which is why the line below is shaded in red. Previously I added them and that seemed to cause problems as well. hmmmm

- Bug in incorrectly allowing construction of :class:`Period` or :class:`PeriodDtype` with :class:`CustomBusinessDay` freq; use :class:`BusinessDay` instead (:issue:`52534`)

File: pandas/tests/io/excel/test_readers.py [hunk 28]
```diff
@@ -148,6 +149,33 @@ def parser(self, *args, **kwargs):
...
+
+        if engine is not None and expected_defaults[read_ext[1:]]:
```
[the remark this answers is not in the document]
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:170]: Not necessary! Fixed!

File: pandas/tests/io/excel/test_readers.py (hunk 26 above)
Comment (rmhowe425) [reply pandas/tests/io/excel/test_readers.py:168]: Fixed!

File: doc/source/whatsnew/v2.1.0.rst (hunk 27 above)
Comment (MarcoGorelli) [reply doc/source/whatsnew/v2.1.0.rst:342]: depends on the merge conflict - if it's a whatsnew note, you probably need to select "keep both changes"

Comment (rmhowe425) [reply doc/source/whatsnew/v2.1.0.rst:342]: Understood! Thank you for the clarity! I'll add the above entry back and keep this in mind moving forward!

**Review by mroeschke — APPROVED** [review]

**Review by MarcoGorelli — APPROVED** [review]
Nice!

Looks like the previous comments have been addressed. This looks like it should be fine, but I'm not too familiar with the Excel code and so would prefer it if someone with more expertise in it were to merge

File: pandas/io/excel/_base.py [hunk 29]
```diff
@@ -302,6 +305,11 @@
 read_csv : Read a comma-separated values (csv) file into DataFrame.
 read_fwf : Read a table of fixed-width formatted lines into DataFrame.
 
+Notes
+-----
+For specific information on the methods used for each Excel engine, refer to the pandas
+user guide: :doc: `/source/user_guide/io`
```
Comment (mroeschke) [inline pandas/io/excel/_base.py:311]:
```suggestion
:ref:`user guide <io.excel_reader>`
```

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Awesome! Thanks @rmhowe425

**Comment by rmhowe425** [comment]
Thanks for all the help guys! Really appreciate it!

**Comment by samukweku** [comment]
hi team, pls what pandas version is this targeted for? I cant find it on pandas 2.0.3

**Comment by rmhowe425** [comment]
@samukweku 2.1.0, which I believe will be released later today. If you check the "Files Changed", you'll see an entry in whatsnew/v2.1.0
